extensions/s3/constructor-s3.go

```
@@ -0,0 +1,110 @@
+// Package s3 provides a treeview.Tree constructor dedicated to AWS S3.
+package s3
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Digital-Shane/treeview"
+	"github.com/Digital-Shane/treeview/extensions/s3/internal/s3"
+)
+
+// NewTreeFromS3 creates a new tree structure based on files fetched from an S3 path, using configurable options.
+// Returns a pointer to a Tree structure or an error if an issue occurs during tree creation.
+//
+// Supported options:
+// Build options:
+//   - treeview.WithFilterFunc:   Filters items during tree building
+//   - treeview.WithMaxDepth:     Limits tree depth during construction
+//   - treeview.WithExpandFunc:   Sets initial expansion state for nodes
+//   - treeview.WithTraversalCap: Limits total nodes processed (returns a partial tree + error if exceeded)
+//   - treeview.WithProgressCallback: Invoked after each filesystem entry is processed (breadth-first per directory)
+func NewTreeFromS3(ctx context.Context, path string, profile string,
+	opts ...treeview.Option[treeview.FileInfo]) (*treeview.Tree[treeview.FileInfo], error) {
+	cfg := treeview.NewMasterConfig(opts, treeview.WithProvider[treeview.FileInfo](treeview.NewDefaultNodeProvider(
+		treeview.WithFileExtensionRules[treeview.FileInfo](),
+	)))
+	nodes, err := buildFileSystemTreeForS3(ctx, path, profile, cfg)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %w", treeview.ErrFileSystem, err)
+	}
+	tree := treeview.NewTreeFromCfg(nodes, cfg)
+	return tree, nil
+}
+
+func buildFileSystemTreeForS3(ctx context.Context, path string, profile string,
+	cfg *treeview.MasterConfig[treeview.FileInfo]) ([]*treeview.Node[treeview.FileInfo], error) {
+	if profile == "" {
+		profile = "default"
```

You should actually allow the empty string to be passed through for profile. You later call WithSharedConfigProfile with this value, and if you pass an empty string to that config option it will be silently ignored, and cause the AWS SDK to use the default provider chain. Which would let the users use any method possible to define the AWS credentials including support for `AWS_PROFILE` and `AWS_DEFAULT_PROFILE`.